Assets/SEE/Layout/NodeLayouts/IncrementalTreeMap/LocalMoves.cs

```diff
@@ -0,0 +1,292 @@
+                        move.Apply();
+                        DeleteNode(obsoleteNode);
+                        return;
+                    }
+                }
+
+                // We should never arrive here
+                Assert.IsFalse(true);
+            }
+        }
+
+        /// <summary>
+        /// Searches the space of layouts, that are similar to the layout of <paramref name="nodes"/>
+        /// (in terms of distance in local moves).
+        /// Apply the layout with the best visual quality to <paramref name="nodes"/>
+        /// </summary>
+        /// <param name="nodes">nodes that represent a layout</param>
+        /// <param name="settings">settings for search</param>
+        public static void LocalMovesSearch(List<Node> nodes, IncrementalTreeMapSetting settings)
+        {
+            var allResults = RecursiveMakeMoves(
+                nodes,
+                new List<LocalMove>(),
+                settings);
+            allResults.Add(new Tuple<List<Node>, double, IList<LocalMove>>
+                (nodes, AspectRatiosPNorm(nodes, settings.PNorm), new List<LocalMove>()));
+            var bestResult = Utils.ArgMin(allResults,
+                x => x.Item2 * 10 + x.Item3.Count).Item1;
+
+            var nodesDictionary = nodes.ToDictionary(n => n.ID, n => n);
+            foreach (var resultNode in bestResult)
+            {
+                nodesDictionary[resultNode.ID].Rectangle = resultNode.Rectangle;
+            }
+
+            Utils.CloneSegments(from: bestResult, to: nodesDictionary);
+        }
+
+        /// <summary>
+        /// Makes recursively local moves on clones of the layout to find similar layout with good visual quality.
+        /// </summary>
+        /// <param name="nodes">nodes that represent a layout</param>
+        /// <param name="movesTillNow">moves that are done before in recursion</param>
+        /// <param name="settings">the settings</param>
+        /// <returns>selection of reached layouts, as tuples of nodes, visual quality measure of the layout and
+        /// the local moves that are applied to get this layout.</returns>
+        private static List<Tuple<List<Node>, double, IList<LocalMove>>> RecursiveMakeMoves(
+            IList<Node> nodes,
+            IList<LocalMove> movesTillNow,
+            IncrementalTreeMapSetting settings)
+        {
+            var resultThisRecursion = new List<Tuple<List<Node>, double, IList<LocalMove>>>();
+            if (movesTillNow.Count >= settings.localMovesDepth) return resultThisRecursion;
+            ICollection<Segment> relevantSegments;
+            if (movesTillNow.Count == 0)
+            {
+                relevantSegments = nodes.SelectMany(n => n.SegmentsDictionary().Values).ToHashSet();
+            }
+            else
+            {
+                var relevantNodes = movesTillNow.SelectMany(m => new[] { m.Node1.ID, m.Node2.ID }).ToHashSet()
+                    .Select(id => nodes.First(n => n.ID == id));
```

Comment:
It should not make any difference, except for performance / complexity where its probably better to make the container unique before lookup every item in linear time. 
But it would not be a big difference in this case, i guess.
However i think it's also more intuitive that the container relevantNodes does not hold any duplicates.
Or would you say it's better to avoid this, because it's not 100% necessary?

Reply:
It does make a difference, I think, because LINQ queries are lazily evaluated.
In this case, it means that the first `.ToHashSet()` (the only non-lazy method in this chain) call causes the sequence to be evaluated at that point, and then evaluated again at the next `.ToHashSet()` call in the next line.

In contrast, when we remove the first `.ToHashSet()`, the sequence will only be evaluated once (i.e., the queries will be combined such that we only iterate once over it). If your aim is to make the sequence items unique at the first `.ToHashSet()` call, you should use the lazy equivalent `.Distinct()` instead of `.ToHashSet()`. This will have the same effect of only keeping unique items at that point without forcing the chain to be evaluated prematurely.

Reply:
Ok, thanks for the explanation, i change it to .Distinct() in 0f0d11ef848cc3ff9c9962fbed455a814387c3f8